File: tests/sample_java/Inefficient.java
/**
 * DeepTrace Sample: Inefficient Java
 * Demonstrates: String concatenation in loop, no-StringBuilder, GC pressure
 */
public class Inefficient {
    public static void main(String[] args) {
        int N = 2000;
        int[] data = new int[N];
        for (int i = 0; i < N; i++) data[i] = N - i;
        
        // Bubble sort
        for (int i = 0; i < N; i++) {
            for (int j = 0; j < N - i - 1; j++) {
                if (data[j] > data[j+1]) { 
                    int t = data[j]; 
                    data[j] = data[j+1]; 
                    data[j+1] = t; 
                }
            }
        }
                
        // String concat in loop — creates many String objects
        String result = "";
        for (int x : data) result += x + ",";
        
        // Create and discard many objects — GC pressure
        for (int i = 0; i < 10000; i++) {
            int[] temp = new int[100];
            for (int j=0; j<100; j++) temp[j] = j * i;
        }
        
        System.out.println("Done: " + result.substring(0, Math.min(50, result.length())));
    }
}

File: tests/sample_python/inefficient.py
"""
DeepTrace Sample: Inefficient Python Code
Demonstrates: O(n²) complexity, excessive memory allocation, GC pressure
"""

def main():
    # Bubble sort on large list (O(n²))
    # Creates many temporary list copies
    # String concatenation in loop (creates many string objects)
    # Repeated list comprehensions that could be generators
    # Deliberately triggers GC pressure with many short-lived objects
    # TARGET METRICS: high allocations, multiple GC events, slow runtime
    
    N = 2000
    data = list(range(N, 0, -1))
    
    # Bubble sort
    for i in range(len(data)):
        for j in range(len(data) - i - 1):
            if data[j] > data[j+1]:
                data[j], data[j+1] = data[j+1], data[j]
                
    # String concatenation in loop
    result = ''
    for x in data[:100]:
        result = result + str(x) + ','
        
    # Memory pressure: create and discard many lists
    for _ in range(1000):
        temp = [x * 2 for x in data]
        del temp
        
    print(f'Sorted first 5: {data[:5]}, result_len: {len(result)}')

if __name__ == "__main__":
    main()
